src/document-submit.js

```diff
@@ -52,11 +52,17 @@ export function onDocumentFormSubmit_(e) {
     return;
   }
 
+  const win = form.ownerDocument.defaultView;
   const action = form.getAttribute('action');
   user().assert(action, 'form action attribute is required: %s', form);
   assertHttpsUrl(action, form, 'action');
   user().assert(!startsWith(action, urls.cdn),
       'form action should not be on AMP CDN: %s', form);
+  // Update the form non-xhr action to add `__amp_source_origin` parameter.
+  // This allows publishers to understand where the request is coming from.
+  if (action.indexOf('__amp_source_origin=') == -1) {
```

Comment:
We need to overwrite the existing `__amp_source_origin`, or fail, to avoid someone trying to pretend to have request from another origin.